### ISSUE TITLE
Improve handling of boolean values in responses

### ISSUE THREAD

**Issue by bbatsov** [nrepl/nrepl#196]
After some recent improvements to the encoding logic now an error is rightfully signalled when a response contains a boolean value. (sidenote - I wonder what was happening with them in the past, as I doubt we managed to filter them out everywhere). 

I've been wondering what to do about booleans in general:

a) nothing - leave it to middleware authors to sanitize the responses
b) convert them to strings (e.g. in bencode or with some middleware dedicated to sanitizing responses). I like the middleware approach more
c) disjoin them from responses, similar to the b) approach
d) come up with some extra type encoding info for bencode? (probably an overkill)
e) something else?

I was reminded about this mostly in the context of the lookup middleware and the fact that I missed some potential booleans that `meta` can return. I'll just add some form of handling this to the particular middleware, but I was wondering if we can do better in general. 

@shen-tian what do you think about this?

**Comment by shen-tian** [comment]
It feels like this is related to the question of "what is the nREPL protocol?" that has come up a few times recently. In that context:

- Does the protocol contain a boolean type?
- If so, how is it encoded in bencode?

Right now, I'd say the answer is "No" and "n/a", but worth questioning. 

Previously, I think it just crashed the bencode transport, and some times silently.

**Comment by bbatsov** [comment]
Yeah, probably you're right. Explicit certainly beats implicit. :-)

> Does the protocol contain a boolean type?

Currently it doesn't.

> If so, how is it encoded in bencode?

N/A

> Right now, I'd say the answer is "No" and "n/a", but worth questioning.

Seems we're like-minded here. :-)

That's not a big deal, but I guess I'll have to at least write some notes about client/extension authors - e.g. use strings instead of booleans if you really need them or something along those lines.

**Comment by shen-tian** [comment]
I'd like to have a go at writing the core nREPL specification at some stage. Just a bit snowed under right now.

**Comment by bbatsov** [comment]
And I'll add here years later that we obviously have a similar problem with requests, as it's hard to represent boolean flags without a boolean type in bencode.
